File: examples/pki/cert_parse_and_print.cpp
#include <chrono>
#include <iostream>
#include <string>
#include <vector>

#include "keylock/crypto/common.hpp"
#include "pki/builder.hpp"
#include "pki/certificate.hpp"
#include "pki/key_utils.hpp"

int main(int argc, char **argv) {
    using namespace std::chrono_literals;

    authbox::pik::Certificate certificate;
    if (argc > 1) {
        auto loaded = authbox::pik::Certificate::load(argv[1]);
        if (!loaded.success) {
            std::cerr << "Failed to load certificate: " << loaded.error << "\n";
            return 1;
        }
        certificate = loaded.value.front();
    } else {
        auto keypair = authbox::pik::generate_ed25519_keypair();
        authbox::pik::CertificateBuilder builder;
        const auto now = std::chrono::system_clock::now();
        builder.set_subject_from_string("CN=On-The-Fly Cert,O=keylock")
            .set_subject_public_key_ed25519(keypair.public_key)
            .set_validity(now - 1h, now + 90 * 24h)
            .set_basic_constraints(false, std::nullopt)
            .set_key_usage(authbox::pik::KeyUsageExtension::DigitalSignature);
        auto cert = builder.build_ed25519(keypair, true);
        if (!cert.success) {
            std::cerr << "Unable to generate fallback certificate: " << cert.error << "\n";
            return 1;
        }
        certificate = cert.value;
    }

    certificate.print_info(std::cout);
    auto sans = certificate.subject_alt_names();
    std::cout << "SubjectAltName count: " << sans.size() << "\n";
    const auto fingerprint = certificate.fingerprint(::keylock::hash::Algorithm::SHA256);
    std::cout << "Fingerprint (SHA-256): " << keylock::crypto::to_hex(fingerprint) << "\n";
    return 0;
}

File: examples/pki/cert_parse_and_print.rs
use authbox::pki::{
    Certificate, CertificateBuilder, DerTime, HashAlgorithm, digest, generate_ed25519_keypair,
    key_usage,
};

fn example_time(year: i32) -> DerTime {
    DerTime {
        year,
        month: 1,
        day: 1,
        hour: 0,
        minute: 0,
        second: 0,
    }
}

fn to_hex(bytes: &[u8]) -> String {
    bytes.iter().map(|byte| format!("{byte:02x}")).collect()
}

fn fallback_certificate() -> Result<Certificate, Box<dyn std::error::Error>> {
    let keypair = generate_ed25519_keypair()?;
    Ok(CertificateBuilder::new()
        .set_subject_from_string("CN=On-The-Fly Cert,O=keylock")?
        .set_subject_public_key_ed25519(keypair.public_key.clone())
        .set_validity(example_time(2024), example_time(2025))
        .set_basic_constraints(false, None)?
        .set_key_usage(key_usage::DIGITAL_SIGNATURE)?
        .build_ed25519_with_self_signed(&keypair, true)?)
}

fn main() -> Result<(), Box<dyn std::error::Error>> {
    let certificate = if let Some(path) = std::env::args().nth(1) {
        Certificate::load_with_relaxed(path, true)?
            .into_iter()
            .next()
            .ok_or("certificate file did not contain a certificate")?
    } else {
        fallback_certificate()?
    };

    print!("{}", certificate.print_info());
    let sans = certificate.subject_alt_names();
    println!("SubjectAltName count: {}", sans.len());
    let fingerprint = digest(HashAlgorithm::Sha256, certificate.der())?;
    println!("Fingerprint (SHA-256): {}", to_hex(&fingerprint));

    Ok(())
}
